js/app/widgets/dynamicLogo.js

```diff
@@ -0,0 +1,308 @@
+// Copyright 2016 Endless Mobile, Inc.
+
+/* exported DynamicLogo */
+
+const EosKnowledgePrivate = imports.gi.EosKnowledgePrivate;
+const Gio = imports.gi.Gio;
+const Gtk = imports.gi.Gtk;
+const GObject = imports.gi.GObject;
+const Rsvg = imports.gi.Rsvg;
+
+const Knowledge = imports.app.knowledge;
+
+const HEIGHT_WIDTH_RATIO = 1.5;
+
+const Width = {
+    MIN: 50,
+    MAX: 400,
+};
+
+const Height = {
+    MIN: 50,
+    MAX: 400,
+};
+
+/**
+ * Class: DynamicLogo
+ *
+ * This widget implements our dynamic logo system. A logo can be composed
+ * of an image, text or both. This widget is considered dynamic because it
+ * can be resized dynamically depending on its allocation size and it can
+ * switch between different layouts depending on the geometry of its
+ * allocation.
+ */
+const DynamicLogo = new Knowledge.Class({
+    Name: 'DynamicLogo',
+    Extends: Gtk.Widget,
+
+    Properties: {
+         /**
+         * Property: image-uri
+         *
+         * A URI to the image displayed in the logo.
+         */
+        'image-uri': GObject.ParamSpec.string('image-uri', 'image-uri', 'image-uri',
+            GObject.ParamFlags.READWRITE | GObject.ParamFlags.CONSTRUCT_ONLY, ''),
+         /**
+         * Property: text
+         *
+         * The text displayed in the logo.
+         */
+        'text': GObject.ParamSpec.string('text', 'text', 'text',
+            GObject.ParamFlags.READWRITE, ''),
+         /**
+         * Property: mode
+         *
+         * The mode of the logo, it can be text, image or full.
+         */
+        'mode': GObject.ParamSpec.string('mode', 'mode', 'mode',
+            GObject.ParamFlags.READWRITE, ''),
+    },
+
+    StyleProperties: {
+        /**
+         * Property: sizing
+         *
+         * -EknDynamicLogo-sizing:
+         *  * 'auto'     - size determined by allocated size. Default.
+         *  * 'size-min' - size determined by min-width and min-height css properties.
+         */
+        'sizing': GObject.ParamSpec.string('sizing', 'sizing', 'sizing',
+            GObject.ParamFlags.READWRITE, 'auto'),
+        /**
+         * Property: text-transform
+         *
+         * -EknDynamicLogo-text-transform:
+         *  * 'none'      - no transformation is performed. Default.
+         *  * 'uppercase' - the text is transformed to uppercase.
+         *  * 'lowercase' - the text is transformed to lowercase.
+         */
+        'text-transform': GObject.ParamSpec.string('text-transform', 'text-transform', 'text-transform',
```

Comment:
Same as above, not a blocker. We should really implement `text-transform` upstream.